lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -925,33 +933,40 @@ findMatchingConcatDimForSameSizeSubTensor(llvm::ArrayRef<size_t> firstDims,
 /// there is a dimension in the input tensors, with which we can meet
 /// two requirements:
 /// 1) Input tensors are concatenate-able along this dimension.
-/// 2) The sub-tensors after this dimension in the input tensors, are
-///    of the same size as the sub-tensor of the input of the original
-///    Concat node after the concatenation dimension.
+/// 2) The trailing/leading dimensions sizes after/before this dimension in the
+///    input tensors, are of the same size as the trailing/leading dimensions of
+///    the input of the original Concat node after/before the concatenation
+///    dimension.
 /// \returns this dimension if found, otherwise -1.
 static int tryToConcatAlongSameSizeSubTensor(llvm::ArrayRef<NodeValue> inputs,
                                              ConcatNode *origConcatN) {
   // For the purpose of the optimiztion
   // Concat(Reshape(X)*N)->Reshape(Concat(N*X)), we want to make sure the new
-  // ConcatNode can concatenate on the sub-tensor which is of the same size of
-  // that of the original Concate node.
+  // ConcatNode can concatenate on the trailing/leading dimensions which are of
+  // the same size of those of the original Concate node.
 
   auto firstDims = inputs.front().dims();
   auto origConcatNInputDims = origConcatN->getInputs().front().dims();
-  // The size of the sub-tensors of the original ConcatNode, which are being
-  // concatenated. This size is simply the product of dimensions following the
-  // dimension used for concatenation.
-  size_t subTensorSizeOriginalConcatNode = 1;
-  for (size_t i = origConcatN->getDim() + 1; i < origConcatNInputDims.size();
-       ++i) {
-    subTensorSizeOriginalConcatNode *= origConcatNInputDims[i];
+  // The sizes of the trailing/leading dimensions of the original ConcatNode,
+  // which are being concatenated. This sizes are simply the products of
+  // dimensions following/before the dimension used for concatenation.
+  size_t trailingDimsProdOriginalConcatNode = 1;
+  size_t leadingDimsProdOriginalConcatNode = 1;
+  for (size_t i = 0; i < origConcatNInputDims.size(); ++i) {
+    if (i < origConcatN->getDim()) {
+      leadingDimsProdOriginalConcatNode *= origConcatNInputDims[i];
+    } else if (i > origConcatN->getDim()) {
+      trailingDimsProdOriginalConcatNode *= origConcatNInputDims[i];
+    }
   }
 
-  // Try to find the dimension in the first input such that the sub-tensor size
-  // based on this direction is the same as the size of sub-tensors based on the
-  // concatenation dimension used by the original ConcatNode.
+  // Try to find the dimension in the first input such that the trailing/leading
+  // dimensions sizes are the same as the sizes of the trailing/leading
+  // dimensions based on the concatenation dimension used by the original
+  // ConcatNode.
   ssize_t dimension = findMatchingConcatDimForSameSizeSubTensor(
```

Comment:
You use "dim" for dimension almost everywhere. May be the variable should be called "dim" as well to be consistent?